### ISSUE TITLE
Associating Metadata with Expressions and Constraints

### ISSUE THREAD

**Issue by ssanderson** [cvxpy/cvxpy#322]
**TL;DR** I'm interested in adding the ability to customize the name of non-variable expressions and/or tag expressions with extra metadata.  I'm curious if you think that's a reasonable potential enhancement for CVXPY.

Hi!  As mentioned over in https://github.com/cvxgrp/cvxpy/issues/313, I'm working on adding a DSL for portfolio optimization to Quantopian/Zipline using CVXPY under the hood.  In general, the structure of my library is that it provides high-level `Objective` and `Constraint` classes like `MaximizeAlpha` or `MaxGrossLeverage`, which expand out into one or many CVXPY Objective/Constraint instances.  For example, here's what a [pair trade](https://en.wikipedia.org/wiki/Pairs_trade) constraint looks like currently:

<details>

```python

class Pair(Constraint):
    """
    A constraint representing a pair of inverse-weighted stocks.

    Parameters
    ----------
    long : Asset
        The asset to long.
    short : Asset
        The asset to short.
    hedge_ratio : float, optional
        The ratio between the respective absolute values of the long and short
        weights. Required to be greater than 0. Default is 1.0, signifying
        equal weighting.
    tolerance : float, optional
        The amount by which the hedge ratio of the calculated weights is
        allowed to differ from the given hedge ratio, in either direction.
        Required to be greater than or equal to 0. Default is 0.0.
    """

    @normalize_positive_number('Pair', ['tolerance'])
    @normalize_strictly_positive_number('Pair', ['hedge_ratio'])
    def __init__(self, long, short, hedge_ratio=1.0, tolerance=0.0):
        self._long = long
        self._short = short
        self._hedge_ratio = hedge_ratio
        self._tolerance = tolerance

    def to_cvxpy(self, old_weights, target_weights):
        # old_weights is a pandas.Series indexed by Asset
        # new_weights is the cvxpy.Variable over which we're optimizing
        universe = old_weights.index

        long_ix, short_ix = universe.get_indexer([self._long, self._short])

        long_weight = target_weights[long_ix]
        short_weight = target_weights[short_ix]

        # The hedge ratio constraint is better understood logically as
        #       hedge_ratio = long_weight / -short_weight,
        #
        # to ensure that hedge_ratio is positive, but rearranged as
        #       hedge_ratio * -short_weight = long_weight,
        #
        # so that we cannot divide by 0. Then we add tolerance to make the
        # upper and lower bound constraints

        max_hedge_ratio = self._hedge_ratio + self._tolerance
        min_hedge_ratio = self._hedge_ratio - self._tolerance

        upper_bound_hedge_ratio_constraint = \
            (long_weight <= -short_weight * max_hedge_ratio)

        lower_bound_hedge_ratio_constraint = \
            (long_weight >= -short_weight * min_hedge_ratio)

        return [
            long_weight >= 0.0,
            short_weight <= 0.0,
            upper_bound_hedge_ratio_constraint,
            lower_bound_hedge_ratio_constraint,
        ]
```
</details>
<br>
An optimization in this higher-level API takes an objective, a list of constraints, and an array of current portfolio weights, all of which gets converted into a CVXPY problem and solved.

I'm starting to focus more on improving error messages and other diagnostic information for when an optimization fails or behaves unexpectedly.  One of the things I'm trying to do is make it easier to figure out which constraints were binding in a particular optimization run.  One of the challenges, I'm finding, is displaying complex CVXPY expressions in a way that's meaningful for a user working at the higher level of abstraction.  In the `Pair` implementation above, for example, the first expression in the returned list corresponds to the constraint that the long stock in the pair has to have a non-negative weight in the final portfolio, but its string representation looks something like `0.0 <= new_weights[300, 0]`, which is confusing, especially if the user specifies more than one pair constraint.

I have two potential ideas for improving this state of affairs:

1. Allow overriding `name` on expression types other than `Variable`.  In my context, this would be useful for making the pair constraint above display as `0.0 <= new_weights['AAPL']` instead of `0.0 <= new_weights[300, 0]`.

2. Provide an API for tagging CVXPY expressions with optional metadata.  I would probably just use this for generating error messages.  I could just implement this without library support by setting a `metadata` attribute on my generated constraints, but that feels a bit hacky.

I'd be curious to hear your thoughts on either of these suggestions.  Would they be welcome and/or straightforward changes?  Is there already some way of doing this that I'm missing?

**Comment by SteveDiamond** [comment]
@ssanderson I like both of your suggestions, and the application you're describing is very interesting. The error messages and printed outputs in cvxpy are quite poor, so I'd certainly appreciate any contribution you make that improves the situation. Making ``name`` a writable property strikes me as a very good idea. 

A metadata/tagging API could also be useful, but I'd need to see a more concrete description of how you would use this to comment further. Having written cvxpy, I don't have a very good perspective on what debugging information normal users need.

**Comment by ssanderson** [comment]
@SteveDiamond I'm not sure a metadata/tagging system would be of much value to a direct user of CVXPY: I'd expect the primary audience for this sort of feature would be people who want to build domain-specific modelling languages on top of CVXPY.  The main use-case I have in mind is for tracking provenance information about what a CVXPY expression represents and/or where it came from.  For example, the main entrypoint to my optimizer essentially looks like this:

```
def calculate_new_weights(objective, constraints, current_weights):
    new_weights = cvxpy.Variable(len(current_weights), name='new_weights')
    cvx_objective = objective.to_cvxpy(current_weights, new_weights)
    constraint_map = {
        c: c.to_cvxpy(current_weights, new_weights)
        for c in all_constraints
    }
    cvx_constraints = list(concat(constraint_map.values()))

    problem = cvxpy.Problem(cvx_objective, cvx_constraints)
    problem.solve()

    if problem.status == cvxpy.OPTIMAL:
        return pd.Series(index=current_weights.index, data=np.array(new_weights.value).ravel())
    else:
        info = run_diagnostics(objective, new_weights, cvx_objective, constraint_map)
        raise OptimizationFailed(info)
```

One of the things that's currently tricky about this setup is that the high-level constraints have most of the semantic information about what the cvxpy constraints represent, but the cvxpy constraints have the information about the specific calculations that result in infeasibility, so I need to join them back up to generate useful diagnostics.  Currently I'm doing this by maintaining and passing around the `constraint_map` dict above, but some of the messaging logic is trickier than I'd like because the high-level constraint has to rediscover what the low-level constraint represents.  (In the pair case it's not so bad because the set of constraints emitted is fixed, but in some cases the number of emitted cvxpy constraints is variable and not obvious to infer.)

**Comment by ssanderson** [comment]
As far as what kinds of debugging tools would be useful to a CVXPY user, here are some of the questions I'm thinking about:

1. When an optimization fails due to infeasibility, which constraints should be relaxed or removed to improve the situation?  This, at least as far as I can see, is a hard problem to solve without domain-specific context, but one reasonably general observation is that there are often natural candidate solutions which are useful to look at for understanding how your constraints interact.  In the portfolio optimization case, for example, two interesting questions are: "Why isn't the empty portfolio feasible?" and "Why isn't the current portfolio feasible?"  A general tool for addressing this might look something like a `propose_solution` method on `cvxpy.Problem`, which would work by initializing the state of problem's inputs and then reporting the feasibility status of the problem's constraints.
2. When an optimization succeeds but produces unexpected results, how can I understand better why the optimization didn't produce the result I expected?  There are a few sub-cases of this I think:
    1. Which constraints were binding on the result of my optimization?  I think this could be accomplished reasonably by providing a method to check whether a constraint is within some tolerance from the boundary of its feasible region given the current values of its parameters.  From what I can see, there isn't currently a straightforward way to express this question in CVXPY (feel free to correct me if I'm wrong about this!).
    2. If I expected a different result from an optimization than what I got, why is the result I got back better than what I expected?  An experience I've had a few times working with CVXPY (mostly while setting up test expectations) is setting up an optimization where I expected a particular result, only to find that the optimization produced a different (and better!) value than what I expected.  CVXPY already provides the ability (though it's not clear to me whether this is public API) to evaluate an expression via its `value` property if the atoms of the expression have been initialized, but I'm not sure if there's a way to provide more information than just the values of the objective at the two points of interest.  In some cases, for example, one could construct a path from the non-optimal point to the optimal point that's monotonically increasing in the value of the objective.  Is that true in all cases? (It's certainly not true in the general optimization case, but I don't have quite as strong an intuition for whether those cases are ruled out by the constraints CVXPY places on problem construction.)

**Comment by ssanderson** [comment]
As far as making `name` a settable property goes, my only worry there would be that changing it from a method to a property would be a breaking change for anyone currently using that method.  Fortunately it's unambiguously breaking: anyone using `name` as a method would get a relatively straigthforward error after upgrading, and the fix is a straightforward find/replace to remove parentheses.

Assuming you're okay with that breakage, I think the most straightforward implementation would be to move `name` to be a `property` on `Expression` and to rename the existing implementations of `name` to something like `default_name`.  The implementation on `Expression` would look something like:

```python
@property
def name(self):
    try:
        return self._user_overridden_name
    except AttributeError:
        return self.default_name()

@name.setter
def name(self, name):
    self._user_overridden_name = name
```
Does that seem reasonable?

**Comment by ssanderson** [comment]
Bump.  @SteveDiamond thoughts on the proposed implementation above?

**Comment by SteveDiamond** [comment]
That's bizarre. I posted a response last week but I don't see it here. My philosophy so far has been to avoid all breaking changes except during major updates. So I would be fine with your name implementation but I want to combine it with a larger set of changes. 

As for the larger metadata API, for the moment why don't you monkey patch in the information you need. Once you have a clear idea of the desired changes, we can merge it into cvxpy.

Responses to your specific points (all good questions!):
1. Constraints have a "constraint.value" field that returns whether they're satisfied. So evaluating points is straightforward. I don't think a generic "propose solution" method makes sense, but it could be part of augmenting cvxpy for a DSL.
2.i. See above.
2. ii. By convexity the line between a non-optimal point and the optimal point will be monotonically non-increasing in the objective (with +\infty for violating constraints). More generally I don't have any particular ideas for analytic tools. Again the best solution is probably domain specific, but I'm open to proposals.

**Comment by ssanderson** [comment]
> That's bizarre. I posted a response last week but I don't see it here.

No worries!  Thanks for re-replying.

> So I would be fine with your name implementation but I want to combine it with a larger set of changes.

Do you think it makes sense for me to submit a patch for this now, or should I wait to include it with other breaking changes?  I see from https://github.com/cvxgrp/cvxpy/issues/199 that there are plans for a CVXPY 1.0 release at some point.  I'm assuming the change I'm proposing would get rolled up in a 1.0 release?

w/r/t my `2.i` section above, I've been able to use `.value` on Constraint instances to check points for violations (this has worked great so far!), but I'm not sure how I would use this to check which constraints were binding on a final (feasible) result.  In the case that the final result comes back optimal or optimal_inaccurate, I'd expect `c.value` to be `True` for all the constraints in my problem.  Is there another attribute on the constraint I should be looking at to check boundary adjacency?  Or, possibly equivalently, is there a straightforward way to find a direction to perturb a point such that I can check if `point + perturbation * epsilon` would violate a given constraint?

**Comment by SteveDiamond** [comment]
Could you submit a patch to the 1.0 branch with your change to name?

Sorry, I misunderstood what you wrote about binding constraints.  You're right, there is no good way to check if a constraint is binding. You could look at the dual variables and see if they're non-zero, but it would be better for there to a be a constraint.binding() boolean function.

**Comment by ssanderson** [comment]
> Could you submit a patch to the 1.0 branch with your change to name?

I'm still planning on getting to this at some point.  I've just been swamped for a while.

**Comment by ClayCampaigne** [comment]
I've implemented a potential solution for constraint/expression naming [here](https://github.com/cvxpy/cvxpy/pull/2928) that's different from the original request, but provides the core functionality without breaking changes. Would something along these lines be of interest, or would the original approach (converting the `name()` method to a property) be preferable? 

  **What's I've implemented:**
- `set_display_name(name)` and `display_name()` methods on expressions and constraints
- Constraints automatically show display names in `str()`: "budget_constraint: sum(x) == 1"
- Expressions keep current math representation in `str()` but show labels via display_name() to prevent contamination in compound expressions
  - Method chaining support: `(x >= 0).set_display_name("non_negative").set_display_name("bounds")`

**Difference from original request**
This approach doesn't overwrite the existing name() method, create a property, or change how mathematical representations are built. Instead, it adds a parallel labeling system. This means:
- ✅no breaking changes
- ✅mathematical expressions stay clean when composed
- ❌Less thoroughgoing than replacing the core naming system
